### PR TITLE
feat: learn git author from apis

### LINKED ISSUE #3332: gitFs learn git author/email via API
Right now you need to configure `gitAuthor` if using `gitFs`, so that Renovate knows which name/email to use for commits. Ideally we should do a "who am I?" query in teh global worker and learn our username/email and only require `gitAuthor` for cases/platforms where such an API call is not supported.

The query could probably be done in lib/worker/global and saved to global config. That way it could still be overwritten on a per-repo basis if desired.

We should research for each platform (github, gitlab, bitbucket cloud, bitbucket server, azure devops) if a suitable API endpoint exists to learn the name/email of the authorized user (i.e. token).

### PR DESCRIPTION
**Pull request by rishabh1403**
Closes #3332

### CODE REVIEW

File: lib/platform/github/index.js [hunk 1]
```diff
@@ -99,7 +100,21 @@ async function getRepos(token, endpoint) {
...
-
+async function getAuthor(token, endpoint) {
+  logger.info('Getting github author from api');
+  const opts = hostRules.find({ platform: 'github' }, { token, endpoint });
+  if (!opts.token) {
+    throw new Error('No token found for getAuthor');
+  }
+  hostRules.update({ ...opts, platform: 'github', default: true });
+  try {
+    const res = await get('user', { paginate: true });
```
Comment (rarkins) [inline lib/platform/github/index.js]: I don't think we need to paginate

Comment (rishabh1403) [reply lib/platform/github/index.js]: Yes, removed that in current commit

**Comment by rishabh1403** [comment]
@rarkins Please review, if all looks good, I will start with tests.

File: lib/workers/global/meta.js [hunk 2]
```diff
@@ -23,5 +25,22 @@ function setMeta(config) {
       name: gitAuthorParsed.name,
       email: gitAuthorParsed.address,
     };
+  } else {
+    const credentials = hostRules.find(config, {});
```
Comment (rarkins) [inline lib/workers/global/meta.js]: Isn't the getAuthor function also doing a hostRules lookup? It seems like we shouldn't need it *twice*

File: lib/platform/bitbucket/index.js [hunk 3]
```diff
@@ -76,6 +77,12 @@ async function getRepos(token, endpoint) {
   }
 }
 
+function getAuthor() {
+  logger.info('Getting bitbucket gitAuthor');
```
Comment (rarkins) [inline lib/platform/bitbucket/index.js]: I think these logger.info messages when we return null would just lead to confusion if users look at logs. Probably best not to log anything at all.

File: lib/workers/global/meta.js [hunk 4]
```diff
@@ -23,5 +25,22 @@ function setMeta(config) {
...
+    try {
+      gitAuthorFromApi = await getPlatformApi(config.platform).getAuthor(
+        credentials.token,
+        credentials.endpoint
+      );
+    } catch (err) {
+      logger.debug('Error getting gitAuthor from api');
+    }
+    if (gitAuthorFromApi && gitAuthorFromApi.name && gitAuthorFromApi.email) {
+      global.gitAuthor = {
```
Comment (rarkins) [inline lib/workers/global/meta.js]: Is this compatible with how we handle a manually configured gitAuthor? I think that we only want to use the autodetect one if there is no gitAuthor provided by the user in config.

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/global/meta.js (hunk 2 above)
Comment (rishabh1403) [reply lib/workers/global/meta.js]: got it :+1:

File: lib/platform/bitbucket/index.js (hunk 3 above)
Comment (rishabh1403) [reply lib/platform/bitbucket/index.js]: makes sense :+1:

File: lib/workers/global/meta.js (hunk 4 above)
Comment (rishabh1403) [reply lib/workers/global/meta.js]: yes, this will only run if `gitAuthor` is absent in config

**Comment by rarkins** [comment]
@rishabh1403 I now think that defaulting to Renovate Bot's own identity for commits is a pretty fine solution if users don't configure their own `gitAuthor`. I think therefore we can close this, however it would be very welcome if you ever get the time/motivation to finish the work and reopen this PR.

**Pull request closed without merging**